src/shims/posix/fs.rs

```diff
@@ -1244,22 +1258,18 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
                 //     pub d_name: [c_char; 256],
                 // }
 
-                let entry_place = this.deref_operand(entry_op)?;
-                let name_place = this.mplace_field(&entry_place, 4)?;
+                let mut name = dir_entry.file_name(); // not a Path as there are no separators!
+                name.push("\0"); // Add a NUL terminator
+                let name_bytes = os_str_to_bytes(&name)?;
+                let name_len = u64::try_from(name_bytes.len()).unwrap();
 
-                let file_name = dir_entry.file_name(); // not a Path as there are no separators!
-                let (name_fits, _) = this.write_os_str_to_c_str(
-                    &file_name,
-                    name_place.ptr,
-                    name_place.layout.size.bytes(),
-                )?;
-                if !name_fits {
-                    throw_unsup_format!(
-                        "a directory entry had a name too large to fit in libc::dirent64"
-                    );
-                }
+                let dirent64_layout = this.libc_ty_layout("dirent64")?;
+                let d_name_offset = dirent64_layout.fields.offset(4 /* d_name */).bytes();
+                let size = d_name_offset.checked_add(name_len).unwrap();
+
+                let entry =
+                    this.malloc(size, /*zero_init:*/ false, MiriMemoryKind::Runtime)?;
 
-                let entry_place = this.deref_operand(entry_op)?;
                 let ino64_t_layout = this.libc_ty_layout("ino64_t")?;
```

Comment:
Hm, now that we have `dirent64_layout` it seems a bit odd to still separately hard-code the field types and order here.

But that is pre-existing, so I am okay with leaving this as-is for now.


```suggestion
                // FIXME: make use of dirent64_layout.
                let ino64_t_layout = this.libc_ty_layout("ino64_t")?;
```

Reply:
Yeah I didn't see a way to easily get the layout of fields from the struct layout.  If there is one I'm happy use it, otherwise I'll add the FIXME

Reply:
I'd say let's go with the FIXME, this PR has waited long enough.

If you want to do a follow-up to clean this up, ping me on Zulip and I can send you some ideas. :)

Reply:
FYI, https://github.com/rust-lang/miri/pull/2021 resolves this FIXME.